mindmaps-dashboard/src/controllers/visualiser.js

```diff
@@ -0,0 +1,319 @@
+                .setOnHoldOnNode(this.holdOnNode);
+
+            engineClient = new EngineClient();
+
+            halParser = new HALParser();
+            halParser.setNewResource((id, p, a, l) => {
+                visualiser.addNode(id, p, a, l)
+            });
+            halParser.setNewRelationship((f, t, l) => {
+                visualiser.addEdge(f, t, l)
+            });
+        },
+
+        attached() {
+            var graph = this.$els.graph;
+
+            function resizeElements() {
+                // set graph div height
+                var divHeight = window.innerHeight - graph.offsetTop - $('.graph-div').offset().top - 20;
+                $('.graph-div').height(divHeight);
+                //set the height of right panel of same size of graph-div
+                $('.properties-tab').height(divHeight+7);
+                //fix the height of panel-body so that it is possible to make it overflow:scroll
+                $('.properties-tab .panel-body').height(divHeight - 85);
+            };
+
+            visualiser.render(graph);
+
+            codeMirror = CodeMirror.fromTextArea(this.$els.graqlEditor, {
+                lineNumbers: true,
+                theme: "dracula",
+                mode: "graql",
+                viewportMargin: Infinity,
+                extraKeys: {
+                    Enter: this.runQuery,
+                    "Shift-Delete": this.clearGraph,
+                    "Shift-Backspace": this.clearGraph
+                }
+            });
+
+            resizeElements();
+            window.onresize = resizeElements;
+
+            $('.properties-tab').hide();
+            var height = window.innerHeight - graph.offsetTop - $('.graph-div').offset().top + 20;
+            // make the list of resources tab resizable with mouse - jQueryUI
+            $('#list-resources-tab').resizable({
+                //make it fixed height and only resizable towards west
+                minHeight: height,
+                maxHeight: height,
+                handles: "w"
+            });
+
+        },
+
+        methods: {
+            /*
+             * User interaction: queries.
+             */
+            runQuery() {
+                const query = codeMirror.getValue();
+
+                // Empty query.
+                if (query == undefined || query.length === 0)
+                    return;
+
+                engineClient.graqlHAL(query, this.graphResponse);
+                this.resetMsg();
+            },
+
+            typeQuery(t, ti) {
+                codeMirror.setValue("match $x " + (t === 'roles' ? 'plays-role' : 'isa') + " " + ti + ";");
+                this.typeInstances = false;
+                this.runQuery();
+            },
+
+            getMetaTypes() {
+                if (this.typeInstances)
+                    this.typeInstances = false;
+                else
+                    engineClient.getMetaTypes(x => {
+                        if (x != null) {
```

Comment:
How can `x` be null ? The server is unreachable ? I ask merely because I can't see how (if the server is unreachable) an error or notification is displayed. In fact in `EngineClient.js` I don't see any error handling. Is it somewhere else ? I kind of would have expected to see it there.